File: zkteco_sync.py
# zkteco_sync.py
import os
import sys
import time
import sqlite3
import logging
from typing import Iterable, List, Tuple, Optional, Dict, Any
from contextlib import contextmanager
from datetime import datetime

from dotenv import load_dotenv
import requests
from zk import ZK, const  # pip install zk

# ---------------- Config ----------------
load_dotenv()

ZK_IP               = os.getenv("ZK_IP", "46.100.10.217")
ZK_PORT             = int(os.getenv("ZK_PORT", "4370"))
ZK_TIMEOUT          = int(os.getenv("ZK_TIMEOUT", "12"))          # seconds
API_URL             = os.getenv("API_URL", "https://portal.sobhe-roshan.ir/api/v1/ElectronicAttendance")
API_PASS            = os.getenv("API_PASS", "1234")               # '' if none
CLEAR_DEVICE_LOGS   = os.getenv("CLEAR_DEVICE_LOGS", "false").lower() == "true"
GET_RETRIES         = int(os.getenv("GET_RETRIES", "3"))
RECONNECT_DELAY_MS  = int(os.getenv("RECONNECT_DELAY_MS", "1200"))
SQLITE_PATH         = os.getenv("SQLITE_PATH", "attendance.db")
TZ                  = os.getenv("TZ", "Asia/Tehran")
LOG_LEVEL           = os.getenv("LOG_LEVEL", "INFO").upper()
BATCH_LIMIT         = int(os.getenv("BATCH_LIMIT", "500"))
HTTP_TIMEOUT        = int(os.getenv("HTTP_TIMEOUT", "15"))

# ---------------- Logging ----------------
logging.basicConfig(
    level=getattr(logging, LOG_LEVEL, logging.INFO),
    format="%(asctime)s %(levelname)s %(message)s",
)
try:
    import pytz  # optional
    os.environ["TZ"] = TZ
except Exception:
    pass

# ---------------- SQLite ----------------
def connect_db(path: str) -> sqlite3.Connection:
    need_init = not os.path.exists(path)
    conn = sqlite3.connect(path, timeout=30, isolation_level=None)  # autocommit off by default with BEGIN
    conn.execute("PRAGMA journal_mode=WAL;")
    conn.execute("PRAGMA busy_timeout=5000;")
    conn.execute("PRAGMA synchronous=NORMAL;")
    if need_init:
        logging.info("Creating SQLite schema at %s", path)
    init_db(conn)
    return conn

def init_db(conn: sqlite3.Connection):
    conn.execute("""
    CREATE TABLE IF NOT EXISTS attendance_logs (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        user_id     TEXT NOT NULL,
        record_time TEXT NOT NULL,
        server_id   TEXT,
        sent_at     TEXT,
        created_at  TEXT NOT NULL DEFAULT (datetime('now')),
        updated_at  TEXT NOT NULL DEFAULT (datetime('now')),
        UNIQUE(user_id, record_time)
    );
    """)
    conn.commit()

def insert_attendance_bulk(conn: sqlite3.Connection, rows: List[Tuple[str, str]]) -> int:
    """
    rows: list of (user_id, record_time_iso)
    INSERT OR IGNORE to avoid duplicates (UNIQUE(user_id, record_time))
    """
    cur = conn.cursor()
    inserted = 0
    cur.execute("BEGIN")
    try:
        for user_id, record_time in rows:
            if not user_id or not record_time:
                continue
            cur.execute(
                "INSERT OR IGNORE INTO attendance_logs (user_id, record_time) VALUES (?, ?)",
                (user_id, record_time)
            )
            if cur.rowcount > 0:
                inserted += 1
        conn.commit()
    except Exception as e:
        conn.rollback()
        logging.exception("DB bulk insert failed: %s", e)
        raise
    return inserted

def fetch_unsent(conn: sqlite3.Connection, limit: int = BATCH_LIMIT) -> List[Tuple[int, str, str]]:
    cur = conn.cursor()
    cur.execute("""
        SELECT id, user_id, record_time
        FROM attendance_logs
        WHERE server_id IS NULL
        ORDER BY record_time ASC
        LIMIT ?
    """, (limit,))
    return cur.fetchall()

def mark_sent(conn: sqlite3.Connection, row_id: int, server_id: str):
    conn.execute("""
        UPDATE attendance_logs
        SET server_id = ?, sent_at = datetime('now'), updated_at = datetime('now')
        WHERE id = ?
    """, (server_id, row_id))
    conn.commit()

# ---------------- ZK helpers ----------------
@contextmanager
def zk_connection(ip: str, port: int, timeout: int):
    zk = ZK(ip=ip, port=port, timeout=timeout, password=0, force_udp=False, ommit_ping=False)
    conn = None
    try:
        conn = zk.connect()
        if conn:
            logging.info("Connected to ZKTeco %s:%s", ip, port)
        yield conn
    finally:
        try:
            if conn:
                conn.disconnect()
        except Exception:
            pass
        logging.info("Disconnected")

def to_iso(dt) -> Optional[str]:
    """
    Convert supported types (datetime/str/int/float) to ISO-8601 string.
    """
    if dt is None:
        return None
    if isinstance(dt, datetime):
        return dt.isoformat()
    if isinstance(dt, (int, float)):
        try:
            return datetime.utcfromtimestamp(float(dt)).isoformat() + "Z"
        except Exception:
            return None
    if isinstance(dt, str):
        s = dt.strip()
        # Try ISO parse
        try:
            # handle trailing Z
            if s.endswith("Z"):
                return datetime.fromisoformat(s.replace("Z", "+00:00")).isoformat()
            return datetime.fromisoformat(s).isoformat()
        except Exception:
            # Fallback: accept as-is
            return s
    return str(dt)

def map_attendance_item(item) -> Optional[Tuple[str, str]]:
    """
    Try to normalize a single attendance record to (user_id, record_time_iso)
    Common keys: user_id/userid/uid/pin/id, timestamp/time/attTime/checkTime
    """
    user_id = None
    record_time = None

    # user id candidates
    for key in ("user_id", "userid", "uid", "user", "enrollNumber", "enroll_number", "pin", "PIN", "id", "UserID"):
        if hasattr(item, key):
            v = getattr(item, key)
            if v is not None:
                user_id = str(v)
                break
        elif isinstance(item, dict) and key in item and item[key] is not None:
            user_id = str(item[key])
            break

    # time candidates
    for key in ("timestamp", "time", "record_time", "recordtime", "attTime", "checkTime"):
        if hasattr(item, key):
            record_time = getattr(item, key)
            break
        elif isinstance(item, dict) and key in item:
            record_time = item[key]
            break

    iso = to_iso(record_time)
    if not user_id or not iso:
        return None
    return user_id, iso

def fetch_logs_once(ip: str, port: int, timeout: int) -> List[Tuple[str, str]]:
    with zk_connection(ip, port, timeout) as conn:
        if not conn:
            logging.warning("Connection failed")
            return []

        try:
            conn.disable_device()
            time.sleep(0.2)
        except Exception:
            pass

        raw_items = []
        try:
            raw_items = conn.get_attendance() or []
        except Exception as e:
            logging.warning("get_attendance error: %s", e)
            raw_items = []

        try:
            conn.enable_device()
        except Exception:
            pass

        results: List[Tuple[str, str]] = []
        for it in raw_items:
            mapped = map_attendance_item(it)
            if mapped:
                results.append(mapped)
        return results

def fetch_logs_with_retries(ip: str, port: int, timeout: int, retries: int, delay_ms: int) -> List[Tuple[str, str]]:
    for i in range(retries):
        logs = fetch_logs_once(ip, port, timeout)
        if logs:
            return logs
        logging.info("No logs returned; retrying (%d/%d)...", i + 1, retries)
        time.sleep(delay_ms / 1000.0)
    return []

def clear_device_logs():
    with zk_connection(ZK_IP, ZK_PORT, ZK_TIMEOUT) as conn:
        if not conn:
            logging.error("Cannot connect to clear logs")
            return
        try:
            if hasattr(conn, "clear_attendance"):
                conn.clear_attendance()
            elif hasattr(conn, "clear_attendance_log"):
                conn.clear_attendance_log()
            else:
                logging.error("No clear_attendance method available")
                return
            logging.info("Device logs cleared.")
        except Exception as e:
            logging.error("Error clearing device logs: %s", e)

# ---------------- API ----------------
DEFAULT_HEADERS = {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "User-Agent": "ZKSync/1.0"
}

def post_attendance(payload: Dict[str, Any],
                    timeout: int = HTTP_TIMEOUT,
                    max_retries: int = 3) -> Optional[str]:
    """
    Send one record to API. Return server_id string on success, None otherwise.
    Retries with simple backoff.
    """
    with requests.Session() as s:
        s.headers.update(DEFAULT_HEADERS)

        for attempt in range(1, max_retries + 1):
            try:
                r = s.post(API_URL, json=payload, timeout=timeout)
                if 200 <= r.status_code < 300:
                    try:
                        data = r.json() if r.content else {}
                    except Exception:
                        data = {}
                    server_id = data.get("res_id") or data.get("id") or (r.text.strip() if r.text else None)
                    if not server_id:
                        logging.warning("No server id in response: %s", r.text[:300])
                        return None
                    return str(server_id)

                if r.status_code == 409:
                    # duplicate, but treat as sent
                    try:
                        data = r.json() if r.content else {}
                    except Exception:
                        data = {}
                    return str(data.get("res_id") or data.get("id") or "DUPLICATE")

                if r.status_code == 422:
                    logging.error("422 Unprocessable: %s", r.text[:500])
                    return None

                # other errors
                logging.warning("API %s: %s", r.status_code, r.text[:300])
            except requests.RequestException as e:
                logging.warning("API error (try %d): %s", attempt, e)

            if attempt < max_retries:
                time.sleep(attempt)  # 1s, 2s, ...
    return None

def send_unsent_rows(conn: sqlite3.Connection, rate_limit_ms: int = 0) -> int:
    rows = fetch_unsent(conn, BATCH_LIMIT)
    sent_count = 0
    for row_id, user_id, record_time in rows:
        payload = {"user_id": user_id, "time": record_time}
        if API_PASS:
            payload["pass"] = API_PASS

        server_id = post_attendance(payload)
        if server_id:
            mark_sent(conn, row_id, server_id)
            sent_count += 1

        if rate_limit_ms > 0:
            time.sleep(rate_limit_ms / 1000.0)
    return sent_count

# ---------------- Main ----------------
def main():
    conn = connect_db(SQLITE_PATH)

    # 1) pull logs from device and insert into DB
    logs = fetch_logs_with_retries(ZK_IP, ZK_PORT, ZK_TIMEOUT, GET_RETRIES, RECONNECT_DELAY_MS)
    if logs:
        inserted = insert_attendance_bulk(conn, logs)
        logging.info("Inserted new rows: %d", inserted)
    else:
        logging.info("No logs to insert.")

    # 2) send unsent
    sent = send_unsent_rows(conn, rate_limit_ms=0)
    logging.info("Sent to server: %d", sent)

    # 3) optionally clear device logs after successful sends
    if CLEAR_DEVICE_LOGS and sent > 0:
        clear_device_logs()

    conn.close()

if __name__ == "__main__":
    # CLI: python zkteco_sync.py [pull|send|both]
    conn = None
    try:
        action = (sys.argv[1] if len(sys.argv) > 1 else "both").lower()
        if action not in {"pull", "send", "both"}:
            print("Usage: python zkteco_sync.py [pull|send|both]")
            sys.exit(2)

        conn = connect_db(SQLITE_PATH)

        if action in {"pull", "both"}:
            logs = fetch_logs_with_retries(ZK_IP, ZK_PORT, ZK_TIMEOUT, GET_RETRIES, RECONNECT_DELAY_MS)
            if logs:
                ins = insert_attendance_bulk(conn, logs)
                logging.info("Inserted new rows: %d", ins)
            else:
                logging.info("No logs to insert.")

        if action in {"send", "both"}:
            sent = send_unsent_rows(conn, rate_limit_ms=0)
            logging.info("Sent to server: %d", sent)

        if CLEAR_DEVICE_LOGS and action in {"both"}:
            # فقط اگر ارسال داشتیم پاک کن—می‌تونی شرط رو تغییر بدی
            # اینجا برای سادگی بدون شرط پاک نمی‌کنیم
            pass

    finally:
        if conn:
            conn.close()

File: zkteco_sync.php
<?php
declare(strict_types=1);
error_reporting(E_ALL);
ini_set('display_errors', '0');

require __DIR__ . '/vendor/autoload.php';

use Jmrashed\Zkteco\Lib\ZKTeco;

date_default_timezone_set('Asia/Tehran');

/* ================== تنظیمات ================== */
const ZK_IP        = '46.100.10.217';
const ZK_PORT      = 4370;

const STORAGE_DIR  = __DIR__ . '/storage';
const SQLITE_PATH  = STORAGE_DIR . '/attendance.sqlite';

const API_URL      = 'https://portal.sobhe-roshan.ir/api/v1/ElectronicAttendance';
const API_PASS     = '1234';      // اگر لازم نیست خالی بگذارید ''
const BATCH_LIMIT  = 500;         // حداکثر ارسال در هر اجرا
const CURL_TIMEOUT = 15;          // ثانیه
/* ============================================ */

/* ====== ابزار DB ====== */
function db(): PDO {
    if (!is_dir(STORAGE_DIR)) {
        if (!mkdir(STORAGE_DIR, 0777, true) && !is_dir(STORAGE_DIR)) {
            throw new RuntimeException('Cannot create storage dir');
        }
    }
    $pdo = new PDO('sqlite:' . SQLITE_PATH);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    ensureSchema($pdo);
    return $pdo;
}

function columnExists(PDO $pdo, string $table, string $column): bool {
    $stmt = $pdo->query("PRAGMA table_info($table)");
    $cols = $stmt ? $stmt->fetchAll(PDO::FETCH_ASSOC) : [];
    foreach ($cols as $c) {
        if (isset($c['name']) && strcasecmp($c['name'], $column) === 0) {
            return true;
        }
    }
    return false;
}

function ensureSchema(PDO $pdo): void {
    // جدول پایه
    $pdo->exec("
        CREATE TABLE IF NOT EXISTS attendance_logs (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            uid        TEXT NOT NULL,
            emp_id     TEXT,
            status     INTEGER,
            ts         TEXT NOT NULL,
            server_id  TEXT NULL,
            raw_json   TEXT,
            created_at TEXT NOT NULL DEFAULT (datetime('now')),
            sent_at    TEXT NULL,
            updated_at TEXT NOT NULL DEFAULT (datetime('now')),
            UNIQUE(uid, ts)
        );
    ");
    $pdo->exec("CREATE UNIQUE INDEX IF NOT EXISTS ux_attendance_uid_ts ON attendance_logs (uid, ts);");

    // برای سازگاری اگر جدول قدیمی بود، ستون‌ها را اضافه کنیم
    foreach ([
        'emp_id' => "ALTER TABLE attendance_logs ADD COLUMN emp_id TEXT;",
        'status' => "ALTER TABLE attendance_logs ADD COLUMN status INTEGER;",
        'server_id' => "ALTER TABLE attendance_logs ADD COLUMN server_id TEXT NULL;",
        'raw_json' => "ALTER TABLE attendance_logs ADD COLUMN raw_json TEXT;",
        'created_at' => "ALTER TABLE attendance_logs ADD COLUMN created_at TEXT NOT NULL DEFAULT (datetime('now'));",
        'sent_at' => "ALTER TABLE attendance_logs ADD COLUMN sent_at TEXT NULL;",
        'updated_at' => "ALTER TABLE attendance_logs ADD COLUMN updated_at TEXT NOT NULL DEFAULT (datetime('now'));",
    ] as $col => $sql) {
        if (!columnExists($pdo, 'attendance_logs', $col)) {
            $pdo->exec($sql);
        }
    }
}

/* ====== دریافت از دستگاه ====== */
function fetchFromDevice(): array {
    $zk = new ZKTeco(ZK_IP, ZK_PORT);

    if (!$zk->connect()) {
        throw new RuntimeException("Cannot connect to device " . ZK_IP . ":" . ZK_PORT);
    }

    // اگر متدهای disable/enable وجود داشته باشند
    try { if (method_exists($zk, 'disableDevice')) { $zk->disableDevice(); } } catch (Throwable $e) {}

    $logs = [];
    try {
        // معمولاً: [ [uid, emp_id, status, datetime], ... ]
        $logs = $zk->getAttendance() ?? [];
    } catch (Throwable $e) {
        // بی‌صدا رد می‌کنیم؛ لااقل قطع ارتباط انجام شود
    }

    try { if (method_exists($zk, 'enableDevice')) { $zk->enableDevice(); } } catch (Throwable $e) {}
    $zk->disconnect();

    return $logs;
}

function mapAttendanceItem($att): ?array {
    // پشتیبانی از آرایه ایندکسی/کلیدی و آبجکت
    $uid = $emp_id = null;
    $status = null;
    $dt = null;

    if (is_array($att)) {
        // ایندکسی
        if (array_key_exists(0, $att)) { $uid    = (string)$att[0]; }
        if (array_key_exists(1, $att)) { $emp_id = (string)$att[1]; }
        if (array_key_exists(2, $att)) { $status = is_numeric($att[2]) ? (int)$att[2] : null; }
        if (array_key_exists(3, $att)) { $dt     = (string)$att[3]; }
        // کلیدی
        if ($uid    === null && isset($att['uid']))        $uid = (string)$att['uid'];
        if ($emp_id === null && isset($att['id']))         $emp_id = (string)$att['id'];
        if ($status === null && isset($att['status']))     $status = is_numeric($att['status']) ? (int)$att['status'] : null;
        if ($dt     === null && isset($att['timestamp']))  $dt = (string)$att['timestamp'];
        if ($dt     === null && isset($att['time']))       $dt = (string)$att['time'];
    } else {
        // آبجکت
        if (isset($att->uid))        $uid = (string)$att->uid;
        if (isset($att->id))         $emp_id = (string)$att->id;
        if (isset($att->status))     $status = is_numeric($att->status) ? (int)$att->status : null;
        if (isset($att->timestamp))  $dt = (string)$att->timestamp;
        if ($dt === null && isset($att->time)) $dt = (string)$att->time;
    }

    if (!$uid || !$dt) return null;

    $t = strtotime($dt);
    $ts = $t !== false ? date('Y-m-d H:i:s', $t) : (string)$dt;

    return [
        'uid'    => $uid,
        'emp_id' => $emp_id ?? '',
        'status' => $status,
        'ts'     => $ts,
        'raw'    => $att,
    ];
}

/* ====== درج در DB با جلوگیری از تکرار ====== */
function insertLogs(PDO $pdo, array $rawLogs): int {
    $stmt = $pdo->prepare("
        INSERT OR IGNORE INTO attendance_logs
            (uid, emp_id, status, ts, server_id, raw_json, created_at)
        VALUES
            (:uid, :emp_id, :status, :ts, NULL, :raw_json, :created_at)
    ");

    $now = date('Y-m-d H:i:s');
    $inserted = 0;

    $pdo->beginTransaction();
    try {
        foreach ($rawLogs as $att) {
            $m = mapAttendanceItem($att);
            if ($m === null) continue;

            $ok = $stmt->execute([
                ':uid'        => $m['uid'],
                ':emp_id'     => $m['emp_id'],
                ':status'     => $m['status'],
                ':ts'         => $m['ts'],
                ':raw_json'   => json_encode($m['raw'], JSON_UNESCAPED_UNICODE),
                ':created_at' => $now,
            ]);
            if ($ok && $stmt->rowCount() === 1) $inserted++;
        }
        $pdo->commit();
    } catch (Throwable $e) {
        $pdo->rollBack();
        throw $e;
    }
    return $inserted;
}

/* ====== ارسال به سرور ====== */
function fetchUnsent(PDO $pdo, int $limit = BATCH_LIMIT): array {
    $stmt = $pdo->prepare("
        SELECT id, uid, ts
        FROM attendance_logs
        WHERE server_id IS NULL
        ORDER BY ts ASC
        LIMIT :lim
    ");
    $stmt->bindValue(':lim', $limit, PDO::PARAM_INT);
    $stmt->execute();
    return $stmt->fetchAll(PDO::FETCH_ASSOC);
}

function markSent(PDO $pdo, int $id, string $serverId): void {
    $stmt = $pdo->prepare("
        UPDATE attendance_logs
        SET server_id = :sid, sent_at = datetime('now'), updated_at = datetime('now')
        WHERE id = :id
    ");
    $stmt->execute([':sid' => $serverId, ':id' => $id]);
}

function postAttendance(array $payload, int $maxRetries = 3, int $timeout = CURL_TIMEOUT): array {
    $headers = [
        'Accept: application/json',
        'Content-Type: application/json',
        'User-Agent: ZKSync-PHP/1.0'
    ];

    for ($attempt = 1; $attempt <= $maxRetries; $attempt++) {
        $ch = curl_init(API_URL);
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_HTTPHEADER     => $headers,
            CURLOPT_POSTFIELDS     => json_encode($payload, JSON_UNESCAPED_UNICODE),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CONNECTTIMEOUT => $timeout,
            CURLOPT_TIMEOUT        => $timeout,
            // اگر SSL self-signed دارید و خطا می‌گیرید (فقط محیط تست):
            // CURLOPT_SSL_VERIFYPEER => false,
            // CURLOPT_SSL_VERIFYHOST => 0,
        ]);
        $body = curl_exec($ch);
        $err  = curl_error($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($err) {
            if ($attempt < $maxRetries) { sleep($attempt); continue; }
            return ['ok'=>false, 'server_id'=>null, 'error'=>"Network error: $err"];
        }

        if ($code >= 200 && $code < 300) {
            $data = json_decode($body ?: "{}", true);
            $sid  = $data['res_id'] ?? $data['id'] ?? ($body ?: null);
            if (!$sid) return ['ok'=>false, 'server_id'=>null, 'error'=>'No server id in response'];
            return ['ok'=>true, 'server_id'=>(string)$sid, 'error'=>null];
        }

        if ($code == 409) { // Duplicate
            $data = json_decode($body ?: "{}", true);
            $sid  = $data['res_id'] ?? $data['id'] ?? 'DUPLICATE';
            return ['ok'=>true, 'server_id'=>(string)$sid, 'error'=>null];
        }

        if ($code == 422) {
            return ['ok'=>false, 'server_id'=>null, 'error'=>"422: ".($body ?: 'Unprocessable')];
        }

        if ($attempt < $maxRetries) { sleep($attempt); continue; }
        return ['ok'=>false, 'server_id'=>null, 'error'=>"$code: $body"];
    }

    return ['ok'=>false, 'server_id'=>null, 'error'=>'Unknown'];
}

function sendUnsentRows(PDO $pdo, int $rateLimitMs = 0): int {
    $rows = fetchUnsent($pdo);
    $sent = 0;

    foreach ($rows as $r) {
        $payload = [
            'user_id' => $r['uid'],
            'time'    => $r['ts'],
        ];
        if (API_PASS !== '') $payload['pass'] = API_PASS;

        $res = postAttendance($payload);
        if ($res['ok'] && $res['server_id']) {
            markSent($pdo, (int)$r['id'], (string)$res['server_id']);
            $sent++;
        } else {
            // برای دیباگ:
            // error_log("Send failed id={$r['id']}: ".$res['error']);
        }

        if ($rateLimitMs > 0) usleep($rateLimitMs * 1000);
    }

    return $sent;
}

/* =============== اجرا =============== */
try {
    $pdo = db();

    // 1) دریافت از دستگاه و ذخیره در DB (بدون تکرار)
    $rawLogs = fetchFromDevice();
    $inserted = insertLogs($pdo, $rawLogs);

    // 2) ارسال موارد ارسال‌نشده به سرور و علامت‌گذاری
    $sent = sendUnsentRows($pdo, 0);

    // 3) خروجی
    header('Content-Type: text/plain; charset=utf-8');
    $total = (int)$pdo->query("SELECT COUNT(*) FROM attendance_logs")->fetchColumn();
    echo "Inserted now: $inserted\nSent to server: $sent\nTotal rows: $total\n";

} catch (Throwable $e) {
    http_response_code(500);
    header('Content-Type: text/plain; charset=utf-8');
    echo "ERROR: " . $e->getMessage() . "\n";
}
